### ISSUE TITLE
Build fails in Ubuntu VM

### ISSUE THREAD

**Issue by jainds** [facebookresearch/ReAgent#254]
When I try to build I get following error: 

```
make -j2
[  4%] Built target gtest
[  8%] Built target gtest_main
[ 12%] Built target gmock
[ 16%] Built target gmock_main
[ 68%] Built target rasp-lib
[ 70%] Linking CXX executable RaspTest
[ 77%] Built target RaspCli
CMakeFiles/RaspTest.dir/reagent/serving/test/DecisionService_test.cpp.o: In function `testing::internal::PrintTo(std::string const&, std::ostream*)':
/home/ai/ReAgent/serving/external/googletest/googletest/include/gtest/gtest-printers.h:509: undefined reference to `testing::internal::PrintStringTo(std::string const&, std::ostream*)'
CMakeFiles/RaspTest.dir/reagent/serving/test/DecisionService_test.cpp.o: In function `testing::AssertionResult testing::internal::CmpHelperEQFailure<unsigned long, unsigned long>(char const*, char const*, unsigned long const&, unsigned long const&)':
/home/ai/ReAgent/serving/external/googletest/googletest/include/gtest/gtest.h:1525: undefined reference to `testing::internal::EqFailure(char const*, char const*, std::string const&, std::string const&, bool)'
CMakeFiles/RaspTest.dir/reagent/serving/test/DecisionService_test.cpp.o: In function `testing::AssertionResult testing::internal::CmpHelperEQFailure<std::string, std::string>(char const*, char const*, std::string const&, std::string const&)':
/home/ai/ReAgent/serving/external/googletest/googletest/include/gtest/gtest.h:1525: undefined reference to `testing::internal::EqFailure(char const*, char const*, std::string const&, std::string const&, bool)'
CMakeFiles/RaspTest.dir/reagent/serving/test/InMemoryLogJoiner_test.cpp.o: In function `reagent::InMemoryLogJoinerTests_Match_Test::TestBody()':
/home/ai/ReAgent/serving/reagent/serving/test/InMemoryLogJoiner_test.cpp:50: undefined reference to `testing::internal::GetBoolAssertionFailureMessage(testing::AssertionResult const&, char const*, char const*, char const*)'
/home/ai/ReAgent/serving/reagent/serving/test/InMemoryLogJoiner_test.cpp:51: undefined reference to `testing::internal::GetBoolAssertionFailureMessage(testing::AssertionResult const&, char const*, char const*, char const*)'
CMakeFiles/RaspTest.dir/reagent/serving/test/InMemoryLogJoiner_test.cpp.o: In function `reagent::InMemoryLogJoinerTests_NotMatch_Test::TestBody()':
/home/ai/ReAgent/serving/reagent/serving/test/InMemoryLogJoiner_test.cpp:99: undefined reference to `testing::internal::GetBoolAssertionFailureMessage(testing::AssertionResult const&, char const*, char const*, char const*)'
/home/ai/ReAgent/serving/reagent/serving/test/InMemoryLogJoiner_test.cpp:100: undefined reference to `testing::internal::GetBoolAssertionFailureMessage(testing::AssertionResult const&, char const*, char const*, char const*)'
CMakeFiles/RaspTest.dir/reagent/serving/test/Ucb_test.cpp.o: In function `reagent::UcbTests_Simple_Test::TestBody()':
/home/ai/ReAgent/serving/reagent/serving/test/Ucb_test.cpp:32: undefined reference to `testing::internal::GetBoolAssertionFailureMessage(testing::AssertionResult const&, char const*, char const*, char const*)'
CMakeFiles/RaspTest.dir/reagent/serving/test/Ucb_test.cpp.o:/home/ai/ReAgent/serving/reagent/serving/test/Ucb_test.cpp:96: more undefined references to `testing::internal::GetBoolAssertionFailureMessage(testing::AssertionResult const&, char const*, char const*, char const*)' follow
CMakeFiles/RaspTest.dir/reagent/serving/test/PlanProvider_test.cpp.o: In function `testing::AssertionResult testing::internal::CmpHelperEQFailure<std::set<std::string, std::less<std::string>, std::allocator<std::string> >, std::set<std::string, std::less<std::string>, std::allocator<std::string> > >(char const*, char const*, std::set<std::string, std::less<std::string>, std::allocator<std::string> > const&, std::set<std::string, std::less<std::string>, std::allocator<std::string> > const&)':
/home/ai/ReAgent/serving/external/googletest/googletest/include/gtest/gtest.h:1525: undefined reference to `testing::internal::EqFailure(char const*, char const*, std::string const&, std::string const&, bool)'
collect2: error: ld returned 1 exit status
CMakeFiles/RaspTest.dir/build.make:305: recipe for target 'RaspTest' failed
make[2]: *** [RaspTest] Error 1
CMakeFiles/Makefile2:148: recipe for target 'CMakeFiles/RaspTest.dir/all' failed
make[1]: *** [CMakeFiles/RaspTest.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
```

my Cmake command had resulted in : 
```
cmake -DCMAKE_PREFIX_PATH=~/libtorch -DCMAKE_CXX_STANDARD=17 ..
-- Boost version: 1.65.1
-- Found the following Boost libraries:
--   system
--   thread
--   filesystem
--   regex
--   context
--   program_options
--   chrono
--   date_time
--   atomic
CMake Warning (dev) at /home/ai/libtorch/share/cmake/Caffe2/public/cuda.cmake:29 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Environment variable CUDA_ROOT is set to:

    /usr/local/cuda

  For compatibility, CMake is ignoring the variable.
Call Stack (most recent call first):
  /home/ai/libtorch/share/cmake/Caffe2/Caffe2Config.cmake:88 (include)
  /home/ai/libtorch/share/cmake/Torch/TorchConfig.cmake:40 (find_package)
  CMakeLists.txt:32 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found CUDA: /usr/local/cuda (found version "10.1")
-- Caffe2: CUDA detected: 10.1
-- Caffe2: CUDA nvcc is: /usr/local/cuda/bin/nvcc
-- Caffe2: CUDA toolkit directory: /usr/local/cuda
-- Caffe2: Header version is: 10.1
-- Found CUDNN: /usr/lib/x86_64-linux-gnu/libcudnn.so
-- Found cuDNN: v7.6.5  (include: /usr/include, library: /usr/lib/x86_64-linux-gnu/libcudnn.so)
-- Autodetected CUDA architecture(s):  6.0
-- Added CUDA NVCC flags for: -gencode;arch=compute_60,code=sm_60
-- Found Torch: /home/ai/libtorch/lib/libtorch.so
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ai/ReAgent/serving/build
```

**Comment by kaiwenw** [comment]
Hi @jainds, our RASP build should be fixed now, could you try again?

**Comment by hongwei-deloitte** [comment]
@kaiwenw,  I got the same error when I built the RASP on Mac OS. You can see the log of cmake and make. Can you help on this? 

`(base) (ReAgent) USNYHHLUAN1:build hluan$ cmake -DCMAKE_PREFIX_PATH=$HOME/libtorch/libtorch -DCMAKE_CXX_STANDARD=17 ..
-- The C compiler identification is AppleClang 12.0.0.12000032
-- The CXX compiler identification is AppleClang 12.0.0.12000032
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /Users/hluan/opt/anaconda3/bin/python (found version "3.7.6")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- Boost version: 1.67.0
-- Found the following Boost libraries:
--   system
--   thread
--   filesystem
--   regex
--   context
--   program_options
--   chrono
--   date_time
--   atomic
-- Found Torch: /Users/hluan/libtorch/libtorch/lib/libtorch.so
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/hluan/Downloads/ReAgent/serving/build
(base) (ReAgent) USNYHHLUAN1:build hluan$ make
Scanning dependencies of target gtest
[  2%] Building CXX object external/googletest/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o
[  4%] Linking CXX static library ../../../lib/libgtest.a
[  4%] Built target gtest
Scanning dependencies of target rasp-lib
[  6%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/DecisionServiceException.cpp.o
[  8%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/ConfigProvider.cpp.o
[ 10%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/DiskConfigProvider.cpp.o
[ 12%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/Containers.cpp.o
[ 14%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/ActionValueScorer.cpp.o
[ 16%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/PytorchActionValueScorer.cpp.o
/Users/hluan/Downloads/ReAgent/serving/reagent/serving/core/PytorchActionValueScorer.cpp:79:1: warning:
      non-void function does not return a value in all control paths
      [-Wreturn-type]
}
^
1 warning generated.
[ 18%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/Operator.cpp.o
[ 20%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/DecisionPlan.cpp.o
[ 22%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/OperatorRunner.cpp.o
[ 25%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/SharedParameterHandler.cpp.o
[ 27%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/OperatorFactory.cpp.o
[ 29%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/LogJoiner.cpp.o
[ 31%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/InMemoryLogJoiner.cpp.o
[ 33%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/RealTimeCounter.cpp.o
[ 35%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/Headers.cpp.o
[ 37%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/LocalRealTimeCounter.cpp.o
[ 39%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/core/DecisionService.cpp.o
[ 41%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/operators/Ucb.cpp.o
[ 43%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/operators/ActionValueScoring.cpp.o
[ 45%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/operators/EpsilonGreedyRanker.cpp.o
[ 47%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/operators/PropensityFit.cpp.o
[ 50%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/operators/Softmax.cpp.o
[ 52%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/operators/SoftmaxRanker.cpp.o
[ 54%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/operators/Frechet.cpp.o
[ 56%] Building CXX object CMakeFiles/rasp-lib.dir/reagent/serving/operators/Expression.cpp.o
[ 56%] Built target rasp-lib
Scanning dependencies of target gtest_main
[ 58%] Building CXX object external/googletest/googletest/CMakeFiles/gtest_main.dir/src/gtest_main.cc.o
[ 60%] Linking CXX static library ../../../lib/libgtest_main.a
[ 60%] Built target gtest_main
Scanning dependencies of target RaspTest
[ 62%] Building CXX object CMakeFiles/RaspTest.dir/reagent/serving/test/DecisionService_test.cpp.o
[ 64%] Building CXX object CMakeFiles/RaspTest.dir/reagent/serving/test/InMemoryLogJoiner_test.cpp.o
[ 66%] Building CXX object CMakeFiles/RaspTest.dir/reagent/serving/test/Frechet_test.cpp.o
[ 68%] Building CXX object CMakeFiles/RaspTest.dir/reagent/serving/test/Ucb_test.cpp.o
[ 70%] Building CXX object CMakeFiles/RaspTest.dir/reagent/serving/test/PlanProvider_test.cpp.o
[ 72%] Building CXX object CMakeFiles/RaspTest.dir/reagent/serving/test/PropensityFit_test.cpp.o
[ 75%] Building CXX object CMakeFiles/RaspTest.dir/reagent/serving/test/Softmax_test.cpp.o
[ 77%] Building CXX object CMakeFiles/RaspTest.dir/reagent/serving/test/SoftmaxRanker_test.cpp.o
[ 79%] Building CXX object CMakeFiles/RaspTest.dir/reagent/serving/test/TestHeaders.cpp.o
[ 81%] Building CXX object CMakeFiles/RaspTest.dir/reagent/serving/test/Expression_test.cpp.o
[ 83%] Building CXX object CMakeFiles/RaspTest.dir/reagent/serving/test/PytorchScoring_test.cpp.o
[ 85%] Linking CXX executable RaspTest
ld: unknown option: --no-as-needed
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [RaspTest] Error 1
make[1]: *** [CMakeFiles/RaspTest.dir/all] Error 2
make: *** [all] Error 2`

**Comment by MisterTea** [comment]
@hongwei-deloitte Looks like clang is missing that flag.  You will need to use gcc or run inside a linux VM/docker

**Comment by hongwei-deloitte** [comment]
@MisterTea, thanks. Will try it.
